checks.d/ceph.py

```diff
@@ -0,0 +1,137 @@
+"""ceph check
+Collects metrics from ceph clusters
+"""
+
+# stdlib
+import os
+import json
+
+# project
+from checks import AgentCheck
+from utils.subprocess_output import get_subprocess_output
+from config import _is_affirmative
+
+NAMESPACE = "ceph"
+
+class Ceph(AgentCheck):
+    """ Collect metrics and events from ceph """
+
+    DEFAULT_CEPH_CMD = '/usr/bin/ceph'
+
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        AgentCheck.__init__(self, name, init_config, agentConfig, instances)
+
+    def _run_command(self, cmd):
+        fd = os.popen(cmd)
+        raw = fd.read()
+        fd.close()
+        return json.loads(raw)
+
+    def _collect_raw(self, ceph_cmd, instance):
+        use_sudo = _is_affirmative(instance.get('use_sudo', False))
+        if use_sudo:
+            test_sudo = os.system('setsid sudo -l < /dev/null')
+            if test_sudo != 0:
+                raise Exception('The dd-agent user does not have sudo access')
+
+        raw = {}
+        for cmd in ('mon_status', 'status', 'df detail', 'osd pool stats', 'osd perf'):
+            try:
+                if use_sudo:
+                    args = ['sudo', ceph_cmd]
+                else:
+                    args = [ceph_cmd]
+                args.extend(cmd.split())
+                args.append('-fjson')
+                output,_,_ = get_subprocess_output(args, self.log)
+                res = json.loads(output)
+            except Exception, e:
+                self.log.warning('Unable to parse data from cmd=%s: %s' % (cmd, str(e)))
+                continue
+
+            name = cmd.replace(' ', '_')
+            raw[name] = res
+
+        return raw
+
+    def _extract_tags(self, raw, instance):
+        tags = instance.get('tags', [])
+        if 'mon_status' in raw:
+            fsid = raw['mon_status']['monmap']['fsid']
+            tags.append(NAMESPACE + '_fsid:%s' % fsid)
+            tags.append(NAMESPACE + '_mon_state:%s' % raw['mon_status']['state'])
+
+        name = instance.get('name', None)
+        if name:
+            tags.append(NAMESPACE + '_name:%s' % name)
+
+        self.tags = tags
+
+    def _extract_metrics(self, raw):
+        if 'osd_perf' in raw:
+            for osdperf in raw['osd_perf']['osd_perf_infos']:
+                tags = self.tags + ['ceph_osd:osd%s' % osdperf['id']]
+                for k,v in osdperf['perf_stats'].iteritems():
```

Comment:
We should whitelist the list of metrics we want to collect just to avoid bad surprises.